File: Interviews/LeetCode/Topic-Wise/Bit-Manipulation/Easy/Set-Mismatch/645.py
class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        xorval, j, setBit, bucket1, bucket2, n = 0, 1, 0, 0, 0, len(nums)
        
        for i in nums:
            xorval ^= i ^ j
            j += 1
            
        for i in range(15):
            if (xorval >> i) & 1:
                setBit = i
                break
                
        j = 1
        for i in nums:
            if (i >> setBit) & 1: bucket1 ^= i
            else: bucket2 ^= i
            if (j >> setBit) & 1: bucket1 ^= j
            else: bucket2 ^= j
            j += 1
        
        for i in nums:
            if i == bucket1:
                return [bucket1, bucket2]
        
        return [bucket2, bucket1]

File: Interviews/LeetCode/Topic-Wise/Bit-Manipulation/Easy/Set-Mismatch/645.cpp
class Solution {
public:
    vector<int> findErrorNums(vector<int>& nums) {
        int xorval = 0, n = nums.size();
        for (int i = 0, j = 1; i < n; i++, j++) {
            xorval ^= nums[i] ^ j;
        }

        int setBit = 0;
        for (int i = 0; i < 15; i++) {
            if ((xorval >> i) & 1) {
                setBit = i;
                break;
            }
        }

        int bucket1 = 0, bucket2 = 0;
        for (int i = 0, j = 1; i < n; i++, j++) {
            if ((nums[i] >> setBit) & 1) {
                bucket1 ^= nums[i];
            } else {
                bucket2 ^= nums[i];
            }

            if ((j >> setBit) & 1) {
                bucket1 ^= j;
            } else {
                bucket2 ^= j;
            }
        }

        for (int i = 0; i < n; i++) {
            if (nums[i] == bucket1) {
                return {bucket1, bucket2};
            }
        }

        return {bucket2, bucket1};
    }
};